test/e2e/tls_util_test.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/operator-framework/operator-sdk/pkg/tlsutil"
+	framework "github.com/operator-framework/operator-sdk/test/e2e/framework"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TestBothAppAndCATLSAssetsExist ensures that when both application
+// and CA TLS assets exist in the k8s cluster for a given cr,
+// the GenerateCert() simply returns those to the caller.
+func TestBothAppAndCATLSAssetsExist(t *testing.T) {
+	f := framework.Global
+	ctx := f.NewTestCtx(t)
+	defer ctx.Cleanup(t)
+	namespace, err := ctx.GetNamespace()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// treat the Pod manifest as the input CRfor the GenerateCert().
+	crKind := "Pod"
+	crName := "example-pod"
+	mCR := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind: crKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      crName,
+			Namespace: namespace,
+		},
+	}
+
+	certName := "app-cert"
+	appSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: tlsutil.ToAppSecretName(crKind, crName, certName),
+		},
+	}
+	appSecret, err = f.KubeClient.CoreV1().Secrets(namespace).Create(appSecret)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	caConfigMapAndSecretName := tlsutil.ToCASecretAndConfigMapName(crKind, crName)
+	caConfigMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: caConfigMapAndSecretName,
+		},
+	}
+	caConfigMap, err = f.KubeClient.CoreV1().ConfigMaps(namespace).Create(caConfigMap)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	caSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: caConfigMapAndSecretName,
+		},
+	}
+	caSecret, err = f.KubeClient.CoreV1().Secrets(namespace).Create(caSecret)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cg := tlsutil.NewSDKCertGenerator(f.KubeClient)
+	ccfg := &tlsutil.CertConfig{
+		CertName: certName,
+	}
+	actualAppSecret, actualCaConfigMap, actualCaSecret, err := cg.GenerateCert(mCR, nil, ccfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(appSecret, actualAppSecret) {
```

Comment:
Can we avoid doing DeepEqual() for the entire objects. That's usually not very helpful as things like resourceVersion might change if the object is updated after creation. Or some label or annotation might be added by the server after creation.

The checks should be as specific as possible. Either check something in the spec to verify equality. Or do it for the UID which uniquely identifies the object to verify they're both the same objects.
```Go
if !reflect.DeepEqual(appSecret.UID, actualAppSecret.UID)
```
or
```Go
if appSecret.UID != actualAppSecret.UID
```
since they're just strings.

Reply:
`That's usually not very helpful as things like resourceVersion might change if the object is updated after creation`
This shouldn't happen because we don't update the secrets and configmap in this test case. I expect the created objects are the exact same as the one returned by GenerateCert().

Reply:
It's okay for this case but my general point is when either reconciling or checking a test condition we should be as specific as possible when checking the equality of two objects.

In this case even if say the resourceVersion changes (for whatever reason) our test should not fail because we just care if it's the same secret object that we created earlier. And that's verifiable by just looking at something unique(like the UID).
Although even the name/namespace uniquely identifies the secret so we don't even need this check strictly speaking.

But you can keep it as `!reflect.DeepEqual(appSecret.UID, actualAppSecret.UID)` for this case.